File: threads/src/main/java/im/threads/internal/adapters/ChatDiffCallback.java
package im.threads.internal.adapters;

import androidx.core.util.ObjectsCompat;
import androidx.recyclerview.widget.DiffUtil;

import java.util.List;

import im.threads.internal.model.ChatItem;

public class ChatDiffCallback extends DiffUtil.Callback {

    private final List<ChatItem> oldList;
    private final List<ChatItem> newList;

    public ChatDiffCallback(List<ChatItem> oldList, List<ChatItem> newList) {
        this.newList = newList;
        this.oldList = oldList;
    }

    @Override
    public int getOldListSize() {
        return oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return oldList.get(oldItemPosition).isTheSameItem(newList.get(newItemPosition));
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return ObjectsCompat.equals(oldList.get(oldItemPosition), newList.get(newItemPosition));
    }
}

File: threads/src/main/java/im/threads/ui/adapters/ChatDiffCallback.kt
package im.threads.ui.adapters

import androidx.core.util.ObjectsCompat
import androidx.recyclerview.widget.DiffUtil
import im.threads.business.models.ChatItem
import im.threads.business.models.ConsultPhrase

class ChatDiffCallback(private val oldList: List<ChatItem>, private val newList: List<ChatItem>) : DiffUtil.Callback() {
    override fun getOldListSize(): Int {
        return oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldList[oldItemPosition].isTheSameItem(newList[newItemPosition])
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldItem = oldList[oldItemPosition]
        val newItem = newList[newItemPosition]
        val isAvatarsVisibilityTheSame = if (oldItem is ConsultPhrase && newItem is ConsultPhrase) {
            oldItem.isAvatarVisible == newItem.isAvatarVisible
        } else {
            true
        }
        return ObjectsCompat.equals(oldList[oldItemPosition], newList[newItemPosition]) && isAvatarsVisibilityTheSame
    }
}
